TechAndCheck/zenodotus
# Stub email confirmation process for applicants

This PR partially implements the email confirmation process for new applicants. The "partial" part is that it does not, in fact, contain _any_ of infrastructure or changes for sending the applicant a confirmation email, but it does:

- Add confirmation-related attributes and methods to the `Applicant` model, using `Devise::Confirmable` as a template (but not actually _using_ that module)
- Add the controller methods and views for guiding the applicant through the process, including actually confirming their application _if_ you know how to magically compose the URL

In the spirit of smaller PRs, I'm submitting this one while filling in the gaps for actually sending the applicant an email.

**Testing the positive path:**
- `rails db:migrate`
- `./bin/dev` (not just `rails s`, because we updated CSS)
- Go to http://localhost:3000/apply/new and submit a new application
- ✅ You should see the ["Next, confirm your email address" screen](https://user-images.githubusercontent.com/4731/188989758-a0add9c3-0358-499f-b606-16a391e8bbfb.png).
- Open the Rails console and generate the confirmation URL for that new applicant `include Rails.application.routes.url_helpers; applicant = Applicant.last; applicant_confirm_url(email: applicant.email, token: applicant.confirmation_token, host: "localhost:3000")`
- Open that URL in a browser
- ✅ You should see the ["Email address confirmed" screen](https://user-images.githubusercontent.com/4731/188990829-549db105-4696-4f72-b050-6e685106a05b.png).

**Testing the error paths:**

There are multiple errors you should attempt to trigger, among them:

- Flip `sent = true` to `sent = false` in [`ApplicantsController#send_confirmation_email`](https://github.com/TechAndCheck/zenodotus/blob/8f72dab31663a4cac383164c51eaf68ee68d779d/app/controllers/applicants_controller.rb#L122) and try submitting a new application. You should see the ["Problem sending confirmation link" message](https://user-images.githubusercontent.com/4731/188992902-449472c0-0a02-468b-99ff-7da36c585c31.png).
- Visit http://localhost:3000/apply/confirm with no parameters (or with junk `email` or `token` params). You should see this we-can-neither-confirm-nor-deny error, ["Email confirmation could not be found"](https://user-images.githubusercontent.com/4731/188992852-29b1439a-5d83-4329-8b8a-310fc11655f4.png).
- Visit the confirmation URL for the applicant you already confirmed. You should see the same "could not be found" message.

Affects #270
